maestro/trainer/models/florence_2/training.py

```diff
@@ -0,0 +1,348 @@
+    make_it_reproducible(avoid_non_deterministic_algorithms=False)
+    training_run_dir = _establish_training_run_dir(
+        training_dir=configuration.training_dir,
+    )
+    configuration = replace(
+        configuration,
+        training_dir=training_run_dir,
+    )
+    checkpoints_leaderboard = CheckpointsLeaderboard(
+        max_checkpoints=configuration.max_checkpoints_to_keep,
+    )
+    processor, model = load_model(
+        model_id_or_path=configuration.model_id_or_path,
+        revision=configuration.revision,
+        device=configuration.device,
+        cache_dir=configuration.transformers_cache_dir,
+    )
+    train_loader, val_loader, test_loader = prepare_data_loaders(
+        dataset_location=configuration.dataset_location,
+        train_batch_size=configuration.train_batch_size,
+        processor=processor,
+        device=configuration.device,
+        num_workers=configuration.loaders_workers,
+        test_loaders_workers=configuration.test_loaders_workers,
+    )
+    if test_loader is None:
+        test_loader = val_loader
+    peft_model = prepare_peft_model(
+        model=model,
+        r=configuration.lora_r,
+        lora_alpha=configuration.lora_alpha,
+        lora_dropout=configuration.lora_dropout,
+        bias=configuration.bias,
+        use_rslora=configuration.use_rslora,
+        init_lora_weights=configuration.init_lora_weights,
+        revision=configuration.revision,
+    )
+    training_metrics_tracker = MetricsTracker.init(metrics=["loss"])
+    validation_metrics_tracker = MetricsTracker.init(metrics=["loss"])
+    run_training_loop(
+        processor=processor,
+        model=peft_model,
+        data_loaders=(train_loader, val_loader),
+        configuration=configuration,
+        checkpoints_leaderboard=checkpoints_leaderboard,
+        training_metrics_tracker=training_metrics_tracker,
+        validation_metrics_tracker=validation_metrics_tracker,
+    )
+    best_model_path = checkpoints_leaderboard.get_best_model()
+    print(f"Loading best model from {best_model_path}")
+    processor, model = load_model(
+        model_id_or_path=best_model_path,
+    )
+    if test_loader is not None:
+        run_validation_epoch(
+            processor=processor,
+            model=model,
+            loader=test_loader,
+            epoch_number=None,
+            configuration=configuration,
+            title="Test",
+        )
+    best_model_dir = os.path.join(configuration.training_dir, "best_model")
+    print(f"Saving best model: {best_model_dir}")
+    model.save_pretrained(best_model_dir)
+    processor.save_pretrained(best_model_dir)
+    summarise_training_metrics(
+        training_metrics_tracker=training_metrics_tracker,
+        validation_metrics_tracker=validation_metrics_tracker,
+        training_dir=configuration.training_dir,
+    )
+    for split_name in ["valid", "test"]:
+        prepare_detection_training_summary(
```

Comment:
We should have separate `train` and `evaluate` functions for every model architecture. I'm happy to talk more about it.

Reply:
As for calculating mAP on `train` and `valid` subsets should also be executed via `MetricsTracker` on every epoch. That's how most object detection libs does it.